Subject: Various spelling fixes
From: Michael Tokarev <mjt@tls.msk.ru>

An assorted set of spelling fixes

diff --git a/disas/nanomips.cpp b/disas/nanomips.cpp
index 90e63b83674..2b096552719 100644
--- a/disas/nanomips.cpp
+++ b/disas/nanomips.cpp
@@ -840 +840 @@ int NMD::Disassemble(const uint16 * data, std::string & dis,
-                                dis = "ASE attribute missmatch";
+                                dis = "ASE attribute mismatch";
diff --git a/hw/misc/trace-events b/hw/misc/trace-events
index b5118acd3fd..155733646d8 100644
--- a/hw/misc/trace-events
+++ b/hw/misc/trace-events
@@ -124 +124 @@ npcm7xx_rng_write(uint64_t offset, uint64_t value, unsigned size) "offset: 0x%04
-stm32f4xx_syscfg_set_irq(int gpio, int line, int level) "Interupt: GPIO: %d, Line: %d; Level: %d"
+stm32f4xx_syscfg_set_irq(int gpio, int line, int level) "Interrupt: GPIO: %d, Line: %d; Level: %d"
diff --git a/hw/net/allwinner-sun8i-emac.c b/hw/net/allwinner-sun8i-emac.c
index 38d328587e3..27b0c37943b 100644
--- a/hw/net/allwinner-sun8i-emac.c
+++ b/hw/net/allwinner-sun8i-emac.c
@@ -582 +582 @@ static uint64_t allwinner_sun8i_emac_read(void *opaque, hwaddr offset,
-    case REG_INT_EN:            /* Interupt Enable */
+    case REG_INT_EN:            /* Interrupt Enable */
diff --git a/hw/ppc/pnv_bmc.c b/hw/ppc/pnv_bmc.c
index 67ebb16c4d5..8fbadaf03d7 100644
--- a/hw/ppc/pnv_bmc.c
+++ b/hw/ppc/pnv_bmc.c
@@ -227 +227 @@ static void hiomap_cmd(IPMIBmcSim *ibs, uint8_t *cmd, unsigned int cmd_len,
-        qemu_log_mask(LOG_GUEST_ERROR, "HIOMAP: unknow command %02X\n", cmd[2]);
+        qemu_log_mask(LOG_GUEST_ERROR, "HIOMAP: unknown command %02X\n", cmd[2]);
diff --git a/hw/usb/ccid-card-emulated.c b/hw/usb/ccid-card-emulated.c
index 2d566f7db10..5c76bed77aa 100644
--- a/hw/usb/ccid-card-emulated.c
+++ b/hw/usb/ccid-card-emulated.c
@@ -304 +304 @@ static void *event_thread(void *arg)
-                        "ERROR: wrong reader: quiting event_thread\n");
+                        "ERROR: wrong reader: quitting event_thread\n");
diff --git a/hw/usb/hcd-ohci.c b/hw/usb/hcd-ohci.c
index f8c64c8b95b..1cf2816772c 100644
--- a/hw/usb/hcd-ohci.c
+++ b/hw/usb/hcd-ohci.c
@@ -1129 +1129 @@ static int ohci_service_td(OHCIState *ohci, struct ohci_ed *ed)
-            /* An error occured so we have to clear the interrupt counter. See
+            /* An error occurred so we have to clear the interrupt counter. See
diff --git a/hw/virtio/vhost.c b/hw/virtio/vhost.c
index 614ccc2bcb6..87fa7b40419 100644
--- a/hw/virtio/vhost.c
+++ b/hw/virtio/vhost.c
@@ -903 +903 @@ check_dev_state:
-        /* An error is occured. */
+        /* An error is occurred. */
diff --git a/include/hw/s390x/css.h b/include/hw/s390x/css.h
index 08c869ab0af..7901ab276ce 100644
--- a/include/hw/s390x/css.h
+++ b/include/hw/s390x/css.h
@@ -136 +136 @@ struct SubchDev {
-    uint16_t migrated_schid; /* used for missmatch detection */
+    uint16_t migrated_schid; /* used for mismatch detection */
diff --git a/qemu-options.hx b/qemu-options.hx
index 104632ea343..fd5e384d98e 100644
--- a/qemu-options.hx
+++ b/qemu-options.hx
@@ -148 +148 @@ SRST
-        there will be one thread per vCPU therefor taking advantage of
+        there will be one thread per vCPU therefore taking advantage of
@@ -2406 +2406 @@ DEF("netdev", HAS_ARG, QEMU_OPTION_netdev,
-    "                use 'poll-us=n' to speciy the maximum number of microseconds that could be\n"
+    "                use 'poll-us=n' to specify the maximum number of microseconds that could be\n"
diff --git a/target/i386/cpu.c b/target/i386/cpu.c
index 5a8c96072e4..9e24e2c4822 100644
--- a/target/i386/cpu.c
+++ b/target/i386/cpu.c
@@ -6916 +6916 @@ static void x86_cpu_get_crash_info_qom(Object *obj, Visitor *v,
-        error_setg(errp, "No crash occured");
+        error_setg(errp, "No crash occurred");
diff --git a/target/i386/machine.c b/target/i386/machine.c
index 233e46bb70b..43a63c6e25d 100644
--- a/target/i386/machine.c
+++ b/target/i386/machine.c
@@ -1159 +1159 @@ static int nested_state_post_load(void *opaque, int version_id)
-        error_report("Recieved unsupported nested state size: "
+        error_report("Received unsupported nested state size: "
diff --git a/target/m68k/op_helper.c b/target/m68k/op_helper.c
index 202498deb51..c9728651f00 100644
--- a/target/m68k/op_helper.c
+++ b/target/m68k/op_helper.c
@@ -120 +120 @@ static const char *m68k_exception_name(int index)
-        return "Unitialized Interruot";
+        return "Uninitialized Interruot";
diff --git a/target/riscv/cpu.c b/target/riscv/cpu.c
index 6a0264fc6b1..d39d560cf47 100644
--- a/target/riscv/cpu.c
+++ b/target/riscv/cpu.c
@@ -478 +478 @@ static void riscv_cpu_realize(DeviceState *dev, Error **errp)
-                qemu_log("vector verison is not specified, "
+                qemu_log("vector version is not specified, "
